src/lib/core/datetime.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+
+#include <assert.h>
+#include <limits.h>
+#include <string.h>
+#include <time.h>
+
+#include "trivia/util.h"
+#include "datetime.h"
+
+/*
+ * Given the seconds from Epoch (1970-01-01) we calculate date
+ * since Rata Die (0001-01-01).
+ * DT_EPOCH_1970_OFFSET is the distance in days from Rata Die to Epoch.
+ */
+static int
+local_dt(int64_t secs)
+{
+	return dt_from_rdn((int)(secs / SECS_PER_DAY) + DT_EPOCH_1970_OFFSET);
+}
+
+static struct tm *
+datetime_to_tm(const struct datetime *date)
+{
+	static struct tm tm;
+
+	memset(&tm, 0, sizeof(tm));
+	int64_t secs = date->epoch;
+	dt_to_struct_tm(local_dt(secs), &tm);
+
+	int seconds_of_day = (int64_t)date->epoch % SECS_PER_DAY;
+	tm.tm_hour = (seconds_of_day / 3600) % 24;
+	tm.tm_min = (seconds_of_day / 60) % 60;
+	tm.tm_sec = seconds_of_day % 60;
+
+	return &tm;
+}
+
+void
+datetime_now(struct datetime *now)
+{
+	struct timeval tv;
+	gettimeofday(&tv, NULL);
+	now->epoch = tv.tv_sec;
+	now->nsec = tv.tv_usec * 1000;
+
+	time_t now_seconds;
+	time(&now_seconds);
+	struct tm tm;
+	localtime_r(&now_seconds, &tm);
+	now->tzoffset = tm.tm_gmtoff / 60;
+}
+
+size_t
+datetime_strftime(char *buf, uint32_t len, const char *fmt,
+		  const struct datetime *date)
+{
+	struct tm *p_tm = datetime_to_tm(date);
+	return strftime(buf, len, fmt, p_tm);
+}
+
+
+/* NB! buf may be NULL, and we should handle it gracefully, returning
+ * calculated length of output string
+ */
+int
+datetime_to_string(char *buf, int len, const struct datetime *date)
+{
+	int offset = date->tzoffset;
+	/* for negative offsets around Epoch date we could get
+	 * negative secs value, which should be attributed to
+	 * 1969-12-31, not 1970-01-01, thus we first shift
+	 * epoch to Rata Die then divide by seconds per day,
```

Comment:
No, negative seconds since epoch are quite ok. Everything will be working and properly displaying. 

[Till some moment when 32-bit value of dates since Rata Die will be underflowed, but this is still larger than we could handle reasonably in the index, and far larger than 0000..9999 range usually accepted by databases]

The comment was rather about problem I've observed before when I divided by `SECS_PERF_DAY` first, then shifted by epoch offset. In this case I've got incorrect 1969-12-31 with bogus timezones, instead of correct  1970-01-01. The point was that order here does matter while crossing Epoch boundary.

Reply:
May be this code is self-explanatory and we could simply remove this confusing comment?

Reply:
I think yes, the code is simple enough. The comment just confuses. Must be either simplified or can be dropped.

Reply:
Done